### PR TITLE
Role event

### PR DESCRIPTION
**Pull request by shargon**
Close https://github.com/neo-project/neo/issues/2410

### CODE REVIEW

File: src/neo/SmartContract/Native/RoleManagement.cs [hunk 1]
```diff
@@ -59,6 +59,7 @@ private void DesignateAsRole(ApplicationEngine engine, Role role, ECPoint[] node
             list.AddRange(nodes);
             list.Sort();
             engine.Snapshot.Add(key, new StorageItem(list));
+            engine.SendNotification(Hash, "designation", new VM.Types.Array(engine.ReferenceCounter, new StackItem[] { (int)role }));
```
Comment (Qiao-Jin) [inline src/neo/SmartContract/Native/RoleManagement.cs:62]: How about adding block `index` into notification?

Comment (shargon) [reply src/neo/SmartContract/Native/RoleManagement.cs:62]: I think that it's and information that it's not added in other events, it's easy to get, what do you think @erikzhang ?

**Comment by erikzhang** [comment]
You need to add it to ABI. And the name of the event should be capitalized.

File: src/neo/SmartContract/Native/RoleManagement.cs (hunk 1 above)
Comment (roman-khimov) [reply src/neo/SmartContract/Native/RoleManagement.cs:62]: Block number really is important for this (to issue proper `getDesignatedByRole` call using notification data only) . And keys would also be convenient to have.

Comment (erikzhang) [reply src/neo/SmartContract/Native/RoleManagement.cs]: It is always the same as the index of the block where the notification is located, right? If so I don't think we need it.

Comment (roman-khimov) [reply src/neo/SmartContract/Native/RoleManagement.cs]: You may be receiving just notification data without any block/transaction metadata, you won't get block index in this case unless it's included in the notification.

File: src/neo/SmartContract/Native/RoleManagement.cs [hunk 2]
```diff
@@ -17,6 +18,28 @@ public sealed class RoleManagement : NativeContract
...
+                    Parameters = new ContractParameterDefinition[]
+                    {
+                        new ContractParameterDefinition()
+                        {
+                            Name = "Role",
+                            Type = ContractParameterType.Integer
+                        },
+                        new ContractParameterDefinition()
+                        {
+                            Name = "BlockNumber",
```
Comment (erikzhang) [inline src/neo/SmartContract/Native/RoleManagement.cs:35]: `BlockIndex`?

File: src/neo/SmartContract/Native/RoleManagement.cs [hunk 3]
```diff
@@ -59,6 +82,7 @@ private void DesignateAsRole(ApplicationEngine engine, Role role, ECPoint[] node
             list.AddRange(nodes);
             list.Sort();
             engine.Snapshot.Add(key, new StorageItem(list));
+            engine.SendNotification(Hash, "Designation", new VM.Types.Array(engine.ReferenceCounter, new StackItem[] { (int)role, engine.PersistingBlock?.Index ?? 0 }));
```
Comment (erikzhang) [inline src/neo/SmartContract/Native/RoleManagement.cs:85]:
```suggestion
            engine.SendNotification(Hash, "Designation", new VM.Types.Array(engine.ReferenceCounter, new StackItem[] { (int)role, engine.PersistingBlock.Index }));
```

I think it is impossible to be `null`.

**Comment by roman-khimov** [comment]
Update `RequiredCallFlags`?

**Review by erikzhang — APPROVED** [review]

**Pull request merged**